synapse/storage/data_stores/main/ui_auth.py

```diff
@@ -0,0 +1,294 @@
+                )
+            except self.db.engine.module.IntegrityError:
+                attempts += 1
+        raise StoreError(500, "Couldn't generate a session ID.")
+
+    async def get_ui_auth_session(self, session_id: str) -> UIAuthSessionData:
+        """Retrieve a UI auth session.
+
+        Args:
+            session_id: The ID of the session.
+        Returns:
+            A dict containing the device information.
+        Raises:
+            SynapseError: if the session is not found.
+        """
+        try:
+            result = await self.db.simple_select_one(
+                table="ui_auth_sessions",
+                keyvalues={"session_id": session_id},
+                retcols=("clientdict", "uri", "method", "description"),
+                desc="get_ui_auth_session",
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: %s" % session_id)
+
+        result["clientdict"] = json.loads(result["clientdict"])
+
+        return UIAuthSessionData(session_id, **result)
+
+    async def mark_ui_auth_stage_complete(
+        self, session_id: str, stage_type: str, result: Union[str, bool, JsonDict],
+    ):
+        """
+        Mark a session stage as completed.
+
+        Args:
+            session_id: The ID of the corresponding session.
+            stage_type: The completed stage type.
+            result: The result of the stage verification.
+        """
+        await self.db.runInteraction(
+            "mark_ui_auth_stage_complete",
+            self._mark_ui_auth_stage_complete,
+            session_id,
+            stage_type,
+            result,
+        )
+
+    def _mark_ui_auth_stage_complete(
+        self, txn, session_id: str, stage_type: str, result: Union[str, bool, JsonDict],
+    ):
+        try:
+            # Get the session to ensure it exists.
+            self.db.simple_select_one_txn(
+                txn,
+                table="ui_auth_sessions",
+                keyvalues={"session_id": session_id},
+                retcols=("session_id",),
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: %s" % session_id)
+
+        # Add (or update) the results of the current stage to the database.
+        self.db.simple_upsert_txn(
+            txn,
+            table="ui_auth_sessions_credentials",
+            keyvalues={"session_id": session_id, "stage_type": stage_type},
+            values={"result": json.dumps(result)},
+        )
+
+    async def get_completed_ui_auth_stages(
+        self, session_id: str
+    ) -> Dict[str, Union[str, bool, JsonDict]]:
+        """
+        Retrieve the completed stages of a UI authentication session.
+
+        Args:
+            session_id: The ID of the session.
+        Returns:
+            The completed stages mapped to the result of the verification of
+            that auth-type.
+        Raises:
+            StoreError: if the session is not found.
```

Comment:
is this right? I think it just returns an empty dict?

Reply:
Yes, it seems to just return empty. Updated the comment!